server/migrations/2020-06-30-135809_remove_mat_views/up.sql

```diff
@@ -0,0 +1,576 @@
+
+
+-- Drop the mviews
+drop view post_mview;
+drop materialized view user_mview;
+drop view community_mview;
+drop materialized view private_message_mview;
+drop view user_mention_mview;
+drop view reply_view;
+drop view comment_mview;
+drop materialized view post_aggregates_mview;
+drop materialized view community_aggregates_mview;
+drop materialized view comment_aggregates_mview;
+
+-- User
+create table user_fast as select * from user_view;
+alter table user_fast add column fast_id serial primary key;
+
+create index idx_user_fast_id on user_fast (id);
+
+drop trigger refresh_user on user_;
+
+create trigger refresh_user
+after insert or update or delete
+on user_
+for each row
+execute procedure refresh_user();
+
+-- Sample insert 
+-- insert into user_(name, password_encrypted) values ('test_name', 'bleh');
+-- Sample delete
+-- delete from user_ where name like 'test_name';
+-- Sample update
+-- update user_ set avatar = 'hai'  where name like 'test_name';
+create or replace function refresh_user()
+returns trigger language plpgsql
+as $$
+begin
+  IF (TG_OP = 'DELETE') THEN
+    delete from user_fast where id = OLD.id;
+  ELSIF (TG_OP = 'UPDATE') THEN
+    delete from user_fast where id = OLD.id;
+    insert into user_fast select * from user_view where id = NEW.id;
+    
+    -- Refresh post_fast, cause of user info changes
+    delete from post_aggregates_fast where creator_id = NEW.id;
+    insert into post_aggregates_fast select * from post_aggregates_view where creator_id = NEW.id;
+
+    delete from comment_aggregates_fast where creator_id = NEW.id;
+    insert into comment_aggregates_fast select * from comment_aggregates_view where creator_id = NEW.id;
+
+  ELSIF (TG_OP = 'INSERT') THEN
+    insert into user_fast select * from user_view where id = NEW.id;
+  END IF;
+
+  return null;
+end $$;
+
+-- Post
+
+create table post_aggregates_fast as select * from post_aggregates_view;
+alter table post_aggregates_fast add column fast_id serial primary key;
+
+create index idx_post_aggregates_fast_id on post_aggregates_fast (id);
+
+-- For the hot rank resorting
+create index idx_post_aggregates_fast_hot_rank_published on post_aggregates_fast (hot_rank desc, published desc);
+
+create view post_fast_view as 
+select
+ap.*,
+u.id as user_id,
+coalesce(pl.score, 0) as my_vote,
+(select cf.id::bool from community_follower cf where u.id = cf.user_id and cf.community_id = ap.community_id) as subscribed,
+(select pr.id::bool from post_read pr where u.id = pr.user_id and pr.post_id = ap.id) as read,
+(select ps.id::bool from post_saved ps where u.id = ps.user_id and ps.post_id = ap.id) as saved
+from user_ u
+cross join (
+  select
+  pa.*
+  from post_aggregates_fast pa
```

Comment:
Agree, this is a really costly join, and I'd be all ears to hear a better way to do it.

So its necessary because outside of the aggregates, every post / comment / community has user specific data that goes along with it (for example, a community has whether your has subscribed to it, a bunch of posts has all your votes, etc.)

If you pass `select * from comment_fast_view where user_id = null`, it'll give you the query w/o that vote / user specific info, but if you pass `user_id = 35`, it'll give you user 35's votes. This is extremely useful on the backend, which receives that user_id for logged in users, and can fetch a bunch of data at once (like a front page of posts, or a different users posts / comments), and all their specific user data comes back with it.

I'd be interested to hear a better way to do it tho, especially one that would preserve the way we fetch posts / comments in batches in many different ways on the back end.

Reply:
It's a bit difficult to provide an alternative because from your description it seems like we have a few ideas that don't play well together and I don't know enough to say how to better accomplish them.  From a purely database pov- a cross join like this will take every row generate via post_aggregates and multiply it by every row from users (which means re-processing any underlying query result that many times) which creates a "god view" than is then potentially filtered by your api and the question that needs asked is _why_ each user needs the aggregated post stats in the same view at the same time when presumably we're fetching for 1 user or no user.  If this has to do with how the backend works the solution probably lies there and this PR may not be the best way to discuss it or the right time to.  Only mention this as I feel if you run this through an explain you'll get a very very high cost which is OK for now but not so much in the future.

Reply:
I think the main problem / reason for the cross-join, is that a user might have no aggregate data for that post, but still needs to see it.

IE, user 35 hasn't voted on this post, or interacted with it in any way.... yet still needs to see it on their front page, when they `select * from post_view where user_id = 35`; So a a join from `user -> post_like -> post` would have no results w/o the cross join.

Its something we can do later if there's a better solution, but I don't know of one.

Reply:
It sounds like a right join could work if the path needs to originate from user, something like `user right join post_likes/post_aggs` would still grab all post results for non-matching users.  I agree it can be done later though, I think there's a few alternatives but it needs a deeper dive on my part and I don't want to step on any ongoing work on the queries atm.